### DISCUSSION TITLE
Can the maintainers share a working contractInstance.createEventFilter example?

### DISCUSSION THREAD

**Discussion by ECBSJ** [wevm/viem#1080] [Question] [answered]
I've been having issues with `publicClient.createContractEventFilter` and `contractInstance.createEventFilter`. None of them are working for me.

Can the maintainers share a working `publicClient.createContractEventFilter` and `contractInstance.createEventFilter` example?

**Comment by jxom** [comment]
Some RPC Providers don't support filters. What RPC Provider are you using?

**Reply by ECBSJ** [reply]
I'm currently using wagmi via WalletConnect's setup with a WalletConnect project ID:

https://docs.walletconnect.com/2.0/web3modal/react/wagmi/installation

Before I would actually query for historical events by using the `queryFilter` method of the ethers `Contract` instance from the `useContract` hook.

Now I am trying to use the `.createEventFilter` method from the `Contract` instance from the `getContract` hook...but it's not working properly.

**Reply by ECBSJ** [reply]
Before I had it working like this:

```
    let array = await myContract.queryFilter("someEventName", undefined, undefined).then().catch(console.error)
```

Now I am attempting this:

```
let filter = await myContract.createEventFilter({
          abi: myContractAbi,
          address: "0x3807DAB03E8519F0F4f4c37568E27a71B138d47b",
          eventName: "someEventName",
    })

let logs = await publicClient.getFilterLogs({ filter })
```

**Reply by ECBSJ** [reply]
I either get an empty array or some other error

**Reply by tmm** [reply]
How do you set up `myContract`? The more info we have (ideally a minimal reproduction), the easier it is to help out.

**Reply by ECBSJ** [reply]
Please check out the code in the below link. I have reproduced it showing how I am trying to `createEventFilter` from the `contractInstance` object.

https://stackblitz.com/edit/vitejs-vite-wp7yea?file=src%2FApp.jsx

I get an error in the console saying that `contractInstance.createEventFilter is not a function`

**Reply by tmm** [reply]
Couple issues in the code you shared.

1. `publicClient` returned from `configureChains` is a function that requires passing in a `chainId`.

```diff
let contractInstance = getContract({
  address: '0x3807DAB03E8519F0F4f4c37568E27a71B138d47b',
  abi: contractAbi,
- publicClient: publicClient,
+ publicClient: publicClient({ chainId: mainnet.id }), // <--- now you have a publicClient on mainnet
});
```

2. Wrong syntax for `createEventFilter`. The event name is chained after `createEventFilter` (e.g. `contract.createEventFilter.endHOURRresults(...)`) instead of adding `'eventName'` property.

```diff
-let filter = await contractInstance.createEventFilter({
+let filter = await contractInstance.createEventFilter.endHOURresults({
  address: '0x3807DAB03E8519F0F4f4c37568E27a71B138d47b',
  abi: contractAbi,
- eventName: 'endHOURresults',
});
```

Here's a version that doesn't have runtime errors: https://stackblitz.com/edit/vitejs-vite-1qnhps?file=src%2FApp.jsx

In the future, would strongly recommend using TypeScript with Viem since it would catch these issues for you.

**Reply by ECBSJ** [reply]
Thank you for that.

But now I am getting an empty array when I should be getting back previous historical events. Even if i put add in `fromBlock` and `toBlock `as parameters, `logs` still returns an empty array...

You can see there definitely is a previous event, `endHOURresults`, from the below list of logs from a previous transaction:

https://etherscan.io/tx/0xbb663ac7ea9f2cd588acf063a53d73d82b35da08ae96a82c6923133a728d0fad#eventlog

**Reply by ECBSJ** [reply]
I'm going to go out on a limb and assume that wagmi/viem does not have functionality to query historical events in the same way that `provider.queryFilter` in ethers does?

**Reply by jxom** [reply, marked as answer]
I'm getting logs fine from Alchemy:

```ts
const publicClient = createPublicClient({
  chain: mainnet,
  transport: http('https://eth-mainnet.alchemyapi.io/v2/...')
})

const contractInstance = getContract({
  address: '0x3807DAB03E8519F0F4f4c37568E27a71B138d47b',
  abi: contractAbi,
  publicClient,
})

const filter = await contractInstance.createEventFilter.endHOURresults({
  fromBlock: 0n,
  toBlock: 'latest',
})
const logs = await getFilterLogs(publicClientMainnet, { filter })
```

**Reply by ECBSJ** [reply]
It's working for me now but not stable. Sometimes the correct logs of events will return and sometimes an error will get thrown below:

```
{
    "error": {
        "code": -32000,
        "message": "filter not found"
    },
    "id": 2,
    "jsonrpc": "2.0"
}
```

Any ideas as to why this happens?

This is how the request payload looks like when this type of error above gets thrown:

```
{
  "jsonrpc": "2.0",
  "id": 2,
  "method": "eth_getFilterLogs",
  "params": [
    "0x9d0e00000000000076d18dc7d3766f4f"
  ]
}
```

**Reply by jxom** [reply]
Ah. That’s because RPC providers dispose of filters after a certain amount of time. For more long-lived queries, you should use `getLogs` (this is what ethers uses internally for queryFilter), or you can recreate the filter. We could add an action for `getLogs` on the contract instance to make this easier with contracts. Will take a look tomorrow.

**Comment by ECBSJ** [comment]
Thank you for the help!
